TESTS/mbed_hal/flash/functional_tests/main.cpp

```diff
@@ -0,0 +1,249 @@
+    ret = flash_erase_sector(&test_flash, erase_sector_boundary);
+    TEST_ASSERT_EQUAL_INT32(0, ret);
+
+    ret = flash_program_page(&test_flash, address, data, test_size);
+    TEST_ASSERT_EQUAL_INT32(0, ret);
+    uint8_t *data_flashed = (uint8_t *)address;
+    TEST_ASSERT_EQUAL_UINT8_ARRAY(data, data_flashed, test_size);
+
+    // sector size might not be same as page size
+    erase_sector_boundary = ALIGN_DOWN(address, flash_get_sector_size(&test_flash, address));
+    ret = flash_erase_sector(&test_flash, erase_sector_boundary);
+    TEST_ASSERT_EQUAL_INT32(0, ret);
+
+    // write another data to be certain we are re-flashing
+    for (uint32_t i = 0; i < test_size; i++) {
+        data[i] = 0xAC;
+    }
+    ret = flash_program_page(&test_flash, address, data, test_size);
+    TEST_ASSERT_EQUAL_INT32(0, ret);
+    TEST_ASSERT_EQUAL_UINT8_ARRAY(data, data_flashed, test_size);
+
+    ret = flash_free(&test_flash);
+    TEST_ASSERT_EQUAL_INT32(0, ret);
+    delete[] data;
+}
+
+// make sure programming works with an unaligned data buffer
+void flash_buffer_alignment_test()
+{
+    flash_t test_flash;
+    int32_t ret = flash_init(&test_flash);
+    TEST_ASSERT_EQUAL_INT32(0, ret);
+
+    const uint32_t page_size = flash_get_page_size(&test_flash);
+    const uint32_t buf_size = page_size + 4;
+    uint8_t *data = new uint8_t[buf_size];
+    for (uint32_t i = 0; i < buf_size; i++) {
+        data[i] = i & 0xFF;
+    }
+
+    // use the last four pages for the alignment test
+    const uint32_t flash_end = flash_get_start_address(&test_flash) + flash_get_size(&test_flash);
+    const uint32_t test_addr = flash_end - page_size * 4;
+    const uint32_t erase_sector_boundary = ALIGN_DOWN(test_addr, flash_get_sector_size(&test_flash, test_addr));
+    erase_range(&test_flash, erase_sector_boundary, flash_end - erase_sector_boundary);
+
+    // make sure page program works with an unaligned data buffer
+    for (uint32_t i = 0; i < 4; i++) {
+        const uint32_t addr = test_addr + i * page_size;
+        ret = flash_program_page(&test_flash, addr, data + i, page_size);
+        TEST_ASSERT_EQUAL_INT32(0, ret);
+        uint8_t *data_flashed = (uint8_t *)addr;
+        TEST_ASSERT_EQUAL_UINT8_ARRAY(data + i, data_flashed, page_size);
+    }
+
+    ret = flash_free(&test_flash);
+    TEST_ASSERT_EQUAL_INT32(0, ret);
+    delete[] data;
+}
+
+// check the execution speed at the start and end of the test to make sure
+// cache settings weren't changed
+void flash_clock_and_cache_test()
+{
+    const int timer_diff_end = time_cpu_cycles(TEST_CYCLES);
+    const int acceptable_range = timer_diff_start / (1000000 / ALLOWED_DRIFT_PPM);
+    TEST_ASSERT_UINT32_WITHIN(acceptable_range, timer_diff_start, timer_diff_end);
+}
+
+utest::v1::status_t greentea_failure_handler(const Case *const source, const failure_t reason) {
+    greentea_case_failure_abort_handler(source, reason);
+    return STATUS_CONTINUE;
```

Comment:
STATUS_CONTINUE shouldn't really be used for the failure handler.
1) because it doesn't actually work properly 
2) if a case fails then the overall test fails and should be reported as such (as per Marcus' recommendations)